### PR TITLE
Submission integration

### PR DESCRIPTION
**Pull request by kamaldodiya**
### Description
This PR integrates different modules needed for submissions endpoint
- Generating .ini file
- Streaming user uploaded data
- Create a workflow in the Seqera Platform
- Returns Workflow ID

### Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-2667

### Review App URL(s) 
None

### Example(s)

```
$ curl 'http://localhost:8013/api/tools/vep/submissions' \
-X 'POST' \     
-H 'Content-Type: multipart/form-data; boundary=----WebKitFormBoundary2RyyLdAFakK7ucEc' \
-H 'Accept: */*' \
--data-binary $'------WebKitFormBoundary2RyyLdAFakK7ucEc\r\nContent-Disposition: form-data; name="genome_id"\r\n\r\na7335667-93e7-11ec-a39d-005056b38ce3\r\n------WebKitFormBoundary2RyyLdAFakK7ucEc\r\nContent-Disposition: form-data; name="input_file"; filename="Venter_chr1.vcf"\r\nContent-Type: text/vcard\r\n\r\n\r\n------WebKitFormBoundary2RyyLdAFakK7ucEc\r\nContent-Disposition: form-data; name="parameters"\r\n\r\n{"transcript_set":"gencode_comprehensive","symbol":false,"biotype":false}\r\n------WebKitFormBoundary2RyyLdAFakK7ucEc--\r\n'  | jq
{
  "submission_id": "2oMMWmZO9pRpZH"
}

# Uploaded files
$ ls -l tmp16ozskvl/
-rw-r--r--  1 kamal  staff    0 12 Jul 12:27 Venter_chr1.vcf
-rw-r--r--  1 kamal  staff  219 12 Jul 12:27 config.ini

# config.ini
$ cat tmp16ozskvl/config.ini 
cache 1
dir_cache /nfs/production/flicek/ensembl/variation/data/VEP/tabixconverted/
species homo_sapiens
assembly GRCh38
cache_version 110
offline 1
force_overwrite 1
symbol 0
biotype 0
transcript_version 1
canonical 1
```

### Checklist

- [x] Black formatting
- [x] Tests

### CODE REVIEW

File: app/vep/models/pipeline_model.py
Comment (Mehrnaz-Charkhchi) [inline app/vep/models/pipeline_model.py]: Do we need a NamedTemporaryFile anymore or can we go with a simple  config.ini file in temp dir?

File: app/vep/vep_resources.py
Comment (veidenberg) [inline app/vep/vep_resources.py]: According to APISpec this should return [JSON with submission ID](https://github.com/Ensembl/ensembl-web-tools-api/blob/cbfea2ef333a491ba9aa66ecece27cc36fdaf889/APISpecification.yaml#L280). We can use a dummy ID until proper workflow submission is implemented in [ENSWBSITES-2656](https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-2656) (in that case update the PR description).

File: app/vep/models/pipeline_model.py
Comment (veidenberg) [inline app/vep/models/pipeline_model.py]: This is not needed. You can use [model_dump_json()](https://docs.pydantic.dev/latest/concepts/serialization/#modelmodel_dump_json) instead.

File: app/vep/models/pipeline_model.py
Comment (veidenberg) [reply app/vep/models/pipeline_model.py]: I'd also prefer simple `config.ini` here.

File: app/vep/utils/nextflow.py
Comment (veidenberg) [inline app/vep/utils/nextflow.py:15]: `pipeline_params.model_dump_json()`

File: app/vep/models/pipeline_model.py
Comment (kamaldodiya) [reply app/vep/models/pipeline_model.py]: @veidenberg  The seqera platform expects stringify encoded json. I tried all combination but couldn't generated it so I have to write that (ugly) code. :(

Comment (veidenberg) [reply app/vep/models/pipeline_model.py]: What about `json.dumps(pipeline_params.model_dump_json())`?

Comment (kamaldodiya) [reply app/vep/models/pipeline_model.py]: This will lead to infinite recursion. I tried all different combination but this is what finally working :(

Comment (veidenberg) [reply app/vep/models/pipeline_model.py]: No it won't, it will just stringify twice, escaping the `"` chars, if that's what nextflow expects.

File: app/vep/models/pipeline_model.py
Comment (azangru) [inline app/vep/models/pipeline_model.py:23]: Is this the preferred way of concatenating/interpolating/templating strings? You don't prefer f-strings?

File: app/vep/models/pipeline_model.py
Comment (azangru) [inline app/vep/models/pipeline_model.py]: Doesn't the method name already tell what the method does? What's the value of this comment?

File: app/core/config.py
Comment (veidenberg) [inline app/core/config.py]: Why in debug mode when not specified?

File: app/vep/models/pipeline_model.py
Comment (azangru) [inline app/vep/models/pipeline_model.py:79]: Out of curiosity and ignorance: is this any better than the tempfile approach?

File: app/report.xml
Comment (veidenberg) [inline app/report.xml]: Add to gitignore?

File: app/vep/vep_resources.py
Comment (Mehrnaz-Charkhchi) [inline app/vep/vep_resources.py:51]: Should we add some checks for ini_parameters.create_config_ini_file() 's result here?

File: app/vep/models/pipeline_model.py
Comment (veidenberg) [reply app/vep/models/pipeline_model.py:79]: Tempfile would just add complexity with no benefit over static filename (random filename, unneeded operations run by tempfile). We are already using tempdir.

File: app/tests/test_blast.py
Comment (veidenberg) [inline app/tests/test_blast.py]: Lots of whitespace noise by Black formatting. Can you do the the formatting in a separate PR (or a commit outside this PR)?

File: app/vep/vep_resources.py
Comment (azangru) [inline app/vep/vep_resources.py]: Nitpicking: I would rename, or question, some things here:

- `stream_obj` -> `streamer` (same name as class, but lower-cased), or `request_streamer`. Otherwise the name is misleading. It is a streamer itself, not a stream.
- `stream_result` -> `has_stream_completed` (to make it obvious that the return value is a boolean). Although I would drop this variable altogether. Looking at the implementation of the stream method, it can only ever return `True`, or it can error out. But you are already handling your exceptions. What's the point in checking this return value as well?
- `parameters` -> `vep_parameters`, or `vep_job_parameters` (there are so many parameters in this function, ugh)

File: app/vep/models/pipeline_model.py
Comment (kamaldodiya) [reply app/vep/models/pipeline_model.py:79]: I think initial assumption was to store in a UPLOAD_DIRECTORY but it make sense to have it in the same directory as vcf data since the directory is unique.

File: app/vep/vep_resources.py
Comment (kamaldodiya) [reply app/vep/vep_resources.py:51]: What sort of check @Mehrnaz-Charkhchi  ?

File: app/tests/test_blast.py
Comment (kamaldodiya) [reply app/tests/test_blast.py]: That is the price I have to pay since this is the first time I did black formatting.

File: app/vep/models/pipeline_model.py
Comment (kamaldodiya) [reply app/vep/models/pipeline_model.py]: I did try all combination @veidenberg  but it's not working.  The only thing which is working is what we have in the code right now. It just needs paramsText as stringified_encoded_json ( with escaped ") hard to achieve with any combination of json, dumps, loads, model_dump, model_dump_json. 
I tried one more time and now getting 400 as the payload is different. 
```
"{\"launch\":{\"computeEnvId\":\"17wuWKrWOZW5JrLoGGwPdD\",\"pipeline\":\"https://github.com/jonk-ebi/platform-test-pipeline\",\"workDir\":\"/homes/ens_adm02/runs\",\"revision\":\"main\",\"pullLatest\":true,\"configProfiles\":[\"slurmnew\"],\"paramsText\":{\"vcf\":\"/Users/kamal/Development/e2020/cicd/e2s/ensembl-web-tools-api/tmpdy1zmjo9/Venter_chr1.vcf\",\"vep_config\":\"/Users/kamal/Development/e2020/cicd/e2s/ensembl-web-tools-api/tmpdy1zmjo9/config.ini\"},\"preRunScript\":\"module load nextflow\",\"postRunScript\":\"\",\"stubRun\":false,\"labelIds\":[],\"headJobCpus\":1,\"headJobMemoryMb\":4}}"
```

Error

```
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.cloud.seqera.io/workflow/launch
```

File: app/core/config.py
Comment (jyothishnt) [inline app/core/config.py]: is this correct?

File: app/tests/test_blast.py
Comment (veidenberg) [reply app/tests/test_blast.py]: That's the point. You don't have to pay the price in this PR and instead commit formatting after PR is merged (or commit to dev branch and rebase). Most of the changes here are now unrelated to this PR. Not very important of course, just the amount of noise makes relevant changes harder to follow (especially in the future).

File: app/core/config.py
Comment (azangru) [reply app/core/config.py]: It'll look from inside the app directory into the data directory that's sitting next to it in the repo; should be correct 

image

File: app/vep/vep_resources.py
Comment (veidenberg) [inline app/vep/vep_resources.py]: This message should be updated since it also catches other errors (e.g. thrown when creating ini file or submitting workflow).

**Review by jyothishnt — APPROVED** [review]
Looks ok to me. Can be merged once other comments are addressed

File: app/core/config.py
Comment (kamaldodiya) [reply app/core/config.py]: did that for local setup/testing shouldn't be this way. Corrected

File: app/vep/models/pipeline_model.py
Comment (kamaldodiya) [reply app/vep/models/pipeline_model.py:23]: With formatting the variable should be enclosed in { } but I need it in the string itself . Also, the way the Pedantic serialiser works this is the only thing which is working I am afraid. I will try to play around with formatting option and see if it works.

File: app/vep/vep_resources.py
Comment (kamaldodiya) [reply app/vep/vep_resources.py]: Done

**Review by azangru — APPROVED** [review]

File: app/tests/test_blast.py
Comment (kamaldodiya) [reply app/tests/test_blast.py]: Well that's jut 2 files from blast. Rest all are rom VEP which very much is being worked on in this PR. Assume this is that PR :) If not then I will remove those two files fro this  PR. :)

Comment (veidenberg) [reply app/tests/test_blast.py]: It's much more that just blast files (see the [commit](https://github.com/Ensembl/ensembl-web-tools-api/commit/3f41f022dc454bcec800a30b96e9175be14076ae)). Moving formatting out of this PR was just a suggestion (for readability), if you wish to keep it here I won't object :)

Comment (kamaldodiya) [reply app/tests/test_blast.py]: @veidenberg  Removed blast  files it now so only changes related to VEP. :)

**Review by Mehrnaz-Charkhchi — APPROVED** [review]

File: app/tests/test_blast.py
Comment (veidenberg) [reply app/tests/test_blast.py]: Removing just the blast files is worst of both worlds: we still have the whitespace noise in this PR and yet have to create a separate commit/PR for whitespace formatting in blast files. Please put back the blast files if you wish to do the formatting in this PR. Thanks!

**Review by veidenberg — APPROVED** [review]

**Pull request merged**
